File: 0314-binary-tree-vertical-order-traversal/BinaryTreeVerticalOrderTraversal.java
import java.util.*;

class Solution {
    
    static class Pair {
        TreeNode node;
        int col;
        
        Pair(TreeNode node, int col) {
            this.node = node;
            this.col = col;
        }
    }
    
    public List<List<Integer>> verticalOrder(TreeNode root) {
        List<List<Integer>> result = new ArrayList<>();
        if (root == null) return result;

        // Column -> List of node values
        Map<Integer, List<Integer>> map = new TreeMap<>();
        Queue<Pair> queue = new LinkedList<>();

        queue.offer(new Pair(root, 0));

        while (!queue.isEmpty()) {
            Pair curr = queue.poll();
            TreeNode node = curr.node;
            int col = curr.col;

            map.putIfAbsent(col, new ArrayList<>());
            map.get(col).add(node.val);

            if (node.left != null) {
                queue.offer(new Pair(node.left, col - 1));
            }
            if (node.right != null) {
                queue.offer(new Pair(node.right, col + 1));
            }
        }

        // Collect results in column order
        for (List<Integer> colList : map.values()) {
            result.add(colList);
        }

        return result;
    }
}

File: 0314-binary-tree-vertical-order-traversal/BinaryTreeVerticalOrderTraversal.cpp
#include <bits/stdc++.h>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    vector<vector<int>> verticalOrder(TreeNode* root) {
        vector<vector<int>> result;
        if (!root) return result;

        // map<column, values>
        map<int, vector<int>> colMap;

        // queue stores node and its column index
        queue<pair<TreeNode*, int>> q;
        q.push({root, 0});

        while (!q.empty()) {
            auto [node, col] = q.front();
            q.pop();

            colMap[col].push_back(node->val);

            if (node->left)
                q.push({node->left, col - 1});
            if (node->right)
                q.push({node->right, col + 1});
        }

        // collect results column by column
        for (auto &it : colMap) {
            result.push_back(it.second);
        }

        return result;
    }
};
